cmd/samConsensus/samConsensus.go

```diff
@@ -19,149 +18,99 @@ import (
+		dna.AllToLower(ref[i].Seq)
 	}
-	err = samFile.Close()
-	exception.PanicOnErr(err)
+	refMap := fasta.ToMap(ref)
 
+	reads, header := sam.GoReadToChan(samFileName)
+	piles := sam.GoPileup(reads, header, false, nil, nil)
 	outVcfFile := fileio.EasyCreate(vcfFile)
 
-	fmt.Fprintf(outVcfFile, "%s\n", strings.Join(vcf.NewHeader(samFileName).Text, "\n"))
-	var current dna.Base
-	var maxList []dna.Base
-	for i := range ref {
-		for k := range ref[i].Seq {
-			current = voter(votingMatrix[ref[i].Name][k], maxList)
-			if current == dna.N && ref[i].Seq[k] != dna.N {
-				current = dna.ToLower(ref[i].Seq[k])
-			} else if current != ref[i].Seq[k] {
-				fmt.Fprintf(outVcfFile, "%s\t%d\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", ref[i].Name, int64(k+1), ".", dna.BaseToString(ref[i].Seq[k]), dna.BaseToString(current), ".", ".", ".", ".")
+	_, err = fmt.Fprintf(outVcfFile, "%s\n", strings.Join(vcf.NewHeader(samFileName).Text, "\n"))
+	exception.PanicOnErr(err)
+
+	var consensusBase, refBase dna.Base
+	var refName string
+	for p := range piles {
+		consensusBase = maxBase(p)
+		if consensusBase == dna.Gap { // original code had commented out gap case
+			if skipGappedConsensus { // ignore gapped position
+				continue
+			} else { // ignore gapped reads
+				p.CountF[dna.Gap] = 0
+				p.CountR[dna.Gap] = 0
+				consensusBase = maxBase(p)
 			}
-			ref[i].Seq[k] = current
+		}
+		if p.CountF[consensusBase] == 0 && p.CountR[consensusBase] == 0 { // skip if no data present
+			continue
+		}
+		refName = header.Chroms[p.RefIdx].Name
+		refBase = dna.ToUpper(refMap[refName][p.Pos-1])
+		refMap[header.Chroms[p.RefIdx].Name][p.Pos-1] = consensusBase
+		if refBase != consensusBase {
+			_, err = fmt.Fprintf(outVcfFile, "%s\t%d\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", refName, int64(p.Pos), ".", dna.BaseToString(refBase), dna.BaseToString(consensusBase), ".", ".", ".", ".")
+			exception.PanicOnErr(err)
 		}
 	}
+
 	err = outVcfFile.Close()
 	exception.PanicOnErr(err)
 
 	fasta.Write(outFile, ref)
 }
 
-func voter(v voteBase, maxList []dna.Base) dna.Base {
-	if v.A+v.G+v.T+v.C+v.Gap == 0 {
-		return dna.N
-	}
-
-	var max int32 = v.A
+func maxBase(p sam.Pile) dna.Base {
+	var max int = p.CountF[dna.A] + p.CountR[dna.A]
 	var outBase dna.Base = dna.A
 
-	if v.G > max {
-		max = v.G
+	if p.CountF[dna.G]+p.CountR[dna.G] > max {
+		max = p.CountF[dna.G] + p.CountR[dna.G]
 		outBase = dna.G
 	}
 
-	if v.C > max {
-		max = v.C
+	if p.CountF[dna.C]+p.CountR[dna.C] > max {
+		max = p.CountF[dna.C] + p.CountR[dna.C]
 		outBase = dna.C
 	}
 
-	if v.T > max {
-		max = v.T
+	if p.CountF[dna.T]+p.CountR[dna.T] > max {
+		max = p.CountF[dna.T] + p.CountR[dna.T]
 		outBase = dna.T
 	}
 
-	if v.Gap > max {
-		max = v.Gap
+	if p.CountF[dna.Gap]+p.CountR[dna.Gap] > max {
+		max = p.CountF[dna.Gap] + p.CountR[dna.Gap]
 		outBase = dna.Gap
 	}
 
 	//now we check for ties
-	var maxCount int32 = 0
-
-	if v.A == max {
-		maxCount++
-		maxList = append(maxList, dna.A)
+	candidateBases := make([]int, 0, 5)
```

Comment:
It is probably a rare case that you need to do this allocation.  One option would be to keep track in the calculations if you have a tie.  If you don't have a tie, then just return the max base, if you do have a tie then go through the following code to build the slice of bases that tie and choose a winner.
Above, tie would start out false.  If you get a tie you mark tie as true.  If you "beat" the current max, then you not only set max, but you also set tie to false.